Checkout adaptions

danimoh opened the pull request:
Adaptions for new hub multi currency checkout features.

danimoh committed:
SignTransaction: extend request type by timeout and fiat information

danimoh committed:
SignTransaction: add PaymentInfoLine

danimoh commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -29,6 +29,22 @@ class SignTransactionApi extends TopLevelApi {
+            parsedRequest.fiatCurrency = this.parseCurrencyInfo(request.fiatCurrency);
+            if ((parsedRequest.fiatAmount === undefined) !== (parsedRequest.fiatCurrency === undefined)) {
+                throw new Errors.InvalidRequestError('fiatAmount and fiatCurrency must be both defined or undefined.');
+            }
```
I could add a check here that `fiatAmount` does not have more decimal digits than allowed by the the currency info.
Not sure though whether that's desirable?

nibhar replied:
I would say it should not throw in case it has more digits than it is supposed to have. However just rounding it to the least significant digit here is maybe not a bad idea.

danimoh replied:
The fiat decimal count is now not part of the API anymore. Rendering and rounding of fiat currency strings is now handled by the browser and this comment does not apply anymore.

danimoh committed:
PaymentInfoLine: remove recipient hover and click

danimoh committed:
SignTransaction: add Timer

danimoh committed:
Timer: remove focus style on mouse leave and remove click styles

danimoh committed:
Timer: expiry tooltip i18n

nibhar commented on client/src/PublicRequest.ts:
```diff
@@ -93,13 +93,23 @@ export type ExportResult = {
     layout?: SignTransactionRequestLayout,
     recipientLabel?: string,
     senderLabel?: string,
+    // TODO these should only get added for 'checkout' layout
```
could be done with something like this:
```typescript
export type SignTransactionRequest = SimpleRequest & TransactionInfo & {
    keyPath: string,
    recipientLabel?: string,
    senderLabel?: string,
} & ({
    layout: 'checkout',
    shopOrigin: string,
    shopLogoUrl?: string,
    time?: number,
    expires?: number,
    fiatCurrency?: CurrencyInfo,
    fiatAmount?: number,
} | { layout?: 'cashlink' | 'standard' });
```

However I think I would prefer to split that type in its components and then expose a union. That would result in a line or two more but would maybe be better readable. 
As this is not longer making use of `SignTransactionRequestLayout` it could potentially be removed, too.

danimoh replied:
I experimented with this too.
However, I was using an enum for `layout` which made it appear as this was not working while actually the problem was that ts couldn't match that ts enum to SignTransactionApi.Layouts on the javascipt side.
I eventually figured this out and managed to map the js enum to the ts enum via
```js
/**
 * @enum {KeyguardRequest.SignTransactionRequestLayout}
 * @readonly
 */
SignTransactionApi.Layouts = {
    STANDARD: /** @type {KeyguardRequest.SignTransactionRequestLayout.STANDARD} */ ('standard'),
    CHECKOUT: /** @type {KeyguardRequest.SignTransactionRequestLayout.CHECKOUT} */ ('checkout'),
    CASHLINK: /** @type {KeyguardRequest.SignTransactionRequestLayout.CASHLINK} */ ('cashlink'),
};
```

However, to not change the existing type for `KeyguardRequest.SignTransactionRequestLayout` I later discarded the change to an enum.
But I kept the separation of the `SignTransactionRequest` into two sub types.
For this also extra handling had to be added to `Parsed` in `Keyguard.d.ts` as we discussed together.

nibhar commented on src/components/Timer.js:
```diff
@@ -0,0 +1,297 @@
+     * @returns {number}
+     */
+    get _progress() {
+        return 1 - this._timeLeft / this._totalTime;
```
`this._totalTime` could be zero. Is the Timer behaving alright if `this._progress = Number.NEGATIVE_INFINITY`?
If additionally `this._timeLeft = 0` it results in  `this._progress = Number.NaN`, too. 
I realise that this is not really something that should happen, but it is not checked anywhere as far as I can tell if the start and end times do actually make sense.

danimoh replied:
`SignTransactionApi` rejects requests with `time >= expires`, thus `_totalTime` will never be 0.
Of course you could argue that `Timer` is a more general component and shouldn't rely on how it is actually used in `SignTransaction`.
However, the timer behaves ok even when `_totalTime === 0`. Note that when `_totalTime === 0` also `_timeLeft === 0` holds true, therefore `_progress` will be `NaN` and the browser will render a full circle.
Also other than in the implementation in `vue-components`, the timer is not reactive on the `startTime` and `endTime`. These values are just set initially on creation. Thus, a Timer instance with `_totalTime === 0` will never serve a useful purpose.

nibhar replied:
I agree, and just wanted to point out, that the timer does not test for its inputs to be valid.

nibhar commented on src/lib/RequestParser.js:
```diff
@@ -256,6 +256,22 @@ class RequestParser { // eslint-disable-line no-unused-vars
+     * @returns {number | undefined}
+     */
+    parseNonNegativeFiniteNumber(value, allowUndefined = true) {
+        if (value === undefined && allowUndefined) {
```
I would swap the conditions
```suggestion
        if (allowUndefined && value === undefined) {
```
as the second condition is only relevant iff the first one is fulfilled.

sisou replied:
The other way around the second one (`allowUndefined`) is also only relevant when the first one is fulfilled :wink:

sisou replied:
IMO it is better as it is.

nibhar replied:
I like it better the other way around, beginning with the boolean that toggles this functionality.

danimoh replied:
I think this doesn't really matter, but I also prefer how it is currently.

nibhar commented on src/request/sign-transaction/SignTransaction.js:
```diff
@@ -102,6 +121,10 @@ class SignTransaction {
         );
+
+        if (request.expires !== undefined) {
+            setTimeout(() => reject(new Errors.RequestExpired()), request.expires - Date.now());
```
This could instead listen for `Timer.Events.EXPIRED`?

danimoh replied:
It could but I would need to forward that through the `PaymentInfoLine` so that I think that the current approach is nicer.

nibhar commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -39,7 +55,7 @@ class SignTransactionApi extends TopLevelApi {
     /**
      * Checks that the given layout is valid
      * @param {any} layout
```
Not in this PR, but this should probably be 
```suggestion
      * @param {unknown} layout
```

danimoh replied:
Why would that be better than `any`?
Isn't `any` appropriate here?
The method can handle `any`thing and returns the parsed layout if the input could actually be parsed as a layout.

nibhar replied:
`unknown` is less permissive than `any`. 
As an example
```ts
(x: any) => { console.log(x.test); }
```
is valid for the typechecker, while
```ts
(x: unknown) => { console.log(x.test); }
```
is not.

Therefore I think it would be better for the parser to assume `unknown` than `any`, generally speaking.

danimoh replied:
I see.
But I'd say, that should be part of a separate PR that makes this change across the whole project.

nibhar replied:
I agree.

nibhar commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -52,6 +68,42 @@ class SignTransactionApi extends TopLevelApi {
 
+    /**
+     * Parses that a currency info is valid.
+     * @param {any} currencyInfo
```
Or maybe not. Did we not have those as `{unknown}`? If not there maybe should be a future PR to make them `{unknown}` across the board.

sisou replied:
There was an effort by Max once to make them unknown, but the Typescript version had too many problem and thus was never merged.

nibhar replied:
I tested it just now and most of them actually work with `{unknown}`.

sisou replied:
Yes, they work. I'm just saying this change was part of a bigger PR which did not get merged because of other problems.

nibhar commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -52,6 +68,42 @@ class SignTransactionApi extends TopLevelApi {
+        }
+
+        // parse currency code
+        if (typeof currencyInfo.code !== 'string'
```
So this will throw for i.e `typeof currencyInfo === 'number'`, but the Error will only show `Invalid currency code undefined`, which I would say is not ideal.

danimoh replied:
This does now not apply anymore as `fiatCurrency` changed from being a currency info object to a simple ISO string.

nibhar commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -66,9 +118,9 @@ class SignTransactionApi extends TopLevelApi {
-    STANDARD: 'standard',
-    CHECKOUT: 'checkout',
-    CASHLINK: 'cashlink',
+    STANDARD: /** @type {'standard'} */ ('standard'),
```
Why is the annotation required?

danimoh replied:
It is an enum over type `KeyguardRequest.SignTransactionRequestLayout` which includes only the constants 'standard', 'checkout' and 'cashlink'. Therefore, the strings need to be casted to these constant values.

danimoh committed:
SignTransaction,PaymentInfoLine: fiatCurrency as ISO 4217 code

Reduce fiatCurrency to an ISO 4217 code string and render as currency with
toLocaleString. This now supports all currency symbols implemented by the
browser and puts the currency symbol before or behind the amount, based on
the user's locale.

danimoh committed:
PublicRequest,Keyguard.d.ts: checkout request type improvements

danimoh committed:
tsconfig: disable incremental builds

Ensure building correct builds from scratch without using cached
compilation results.

nibhar commented on client/src/PublicRequest.ts:
```diff
@@ -93,15 +93,25 @@ export type ExportResult = {
 };
 
-export type SignTransactionRequest = SimpleRequest & TransactionInfo & {
+export type SignTransactionRequestSimple = SimpleRequest & TransactionInfo & {
```
I don't know if I like `Simple` here as it still is somewhat of a complex request in comparison.

Aside from that, nice work on the types.

danimoh replied:
Suggestions?

nibhar commented on src/components/PaymentInfoLine.js:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * @param {object}           paymentInfo
+     * @param {string}           paymentInfo.recipient
+     * @param {string|undefined} [paymentInfo.label]
```
I think the `| undefined` are not required, as the brackets in `[paymentInfo.label]` already take care of that. Same for the other optional fields.

danimoh replied:
Unfortunately typescript doesn't understand that sub properties of an object in this notation that are put into square brackets are optional, therefore the need to explicitly specify `undefined`.

nibhar commented on src/components/Timer.js:
```diff
@@ -0,0 +1,297 @@
+        // animate via JS.
+        /** @private
+         *  @type {Tweenable} */
+        this._radius = new Tweenable(8);
```
That `8` could be a variable, or a constant. The `12` later on (`_showDetails()`) could be `1.5 * this.timerRadius`, or `1.5 * Timer.RADIUS` respectively.

sisou commented on src/components/PaymentInfoLine.js:
```diff
@@ -0,0 +1,78 @@
+     * @param {number|undefined} [paymentInfo.fiatAmount]
+     * @param {string|undefined} [paymentInfo.fiatCurrency]
+     * @param {number|undefined} [paymentInfo.time]
+     * @param {number|undefined} [paymentInfo.expires]
```
That's cool that you can do object prop types like this! :open_mouth: 
However, wouldn't it be better to define the `PaymentInfo` interface as a type and simply use the type here?

sisou replied:
I still think a `@type` definition makes sense for this instead.

danimoh replied:
Changed it to a `@typedef`

nibhar commented on src/components/Timer.js:
```diff
@@ -0,0 +1,297 @@
+     * @param {boolean} [includeUnit=true]
+     * @returns {string|number}
+     */
+    static _toSimplifiedTime(millis, includeUnit = true) {
```
This function is private and only ever used on `this.timeLeft`.
Is there a reason to have it static instead of
```js
_toSimplifiedTime(includeUnit = true) {...}
```
Additionally I think `_toSimplifiedTime` is maybe not descriptive for what this does. Maybe something along the lines of `_toSmallestPossibleTimeUnit()`?

It might also make sense to have the return type as `string` only, as it is either cast to `string` or `toString`ed anyways.

sisou replied:
While I agree on making it non-static and the return type (string), I think the name is alright. `_toSmallestPossibleTimeUnit` would suggest to me that it's always in seconds (or milliseconds), which is not the case.

Maybe `_toRoundedTime`?

danimoh replied:
I think from the suggested names, `_toSimplifiedTime` is still my favorite.

Why do you think that non-static here would be nicer?

But I agree on the return type.

nibhar replied:
As I see it, it is not really static the way it is used. It is always used on `this.timeLeft`. It is also private so it cannot get used anywhere else either.  I would prefer it to be `get _simplifiedTimeLeft()` but the includeUnit argument prevents that.

Therefore I would have it like this
```js
_toSimplifiedTime(includeUnit = true) {
        // find appropriate unit, starting with second
        let resultTime = this._timeLeft / 1000;
...
```

sisou replied:
I'm with Sebastian. If not used statically and only used from an instance, it should not be static and just be on the instance.

nibhar commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -23,14 +24,29 @@ class SignTransactionApi extends TopLevelApi {
         parsedRequest.layout = this.parseLayout(request.layout);
-        if (parsedRequest.layout === SignTransactionApi.Layouts.CHECKOUT) {
+        if (parsedRequest.layout === SignTransactionApi.Layouts.CHECKOUT
+            && request.layout === SignTransactionApi.Layouts.CHECKOUT) {
```
Why do we need to check the layout for both the parsed and unparsed request?

sisou replied:
I have the same question.

danimoh replied:
It's required for ts to be able to infer that the checkout specific properties exist on `request` and `parsedRequest`.

nibhar reviewed (approved):
Approved with some remarks.

sisou commented on src/components/PaymentInfoLine.js:
```diff
@@ -0,0 +1,78 @@
+     * @param {string}           paymentInfo.recipient
+     * @param {string|undefined} [paymentInfo.label]
+     * @param {URL|undefined}    [paymentInfo.imageUrl]
+     * @param {number}           paymentInfo.lunaAmount
```
Does it really have to be called `lunaAmount`? I feel the default for `amount` should just be luna by definition.

danimoh replied:
I guess I put it that way for easier distinction from the fiat amount.

sisou commented on src/lib/NumberFormatting.js:
```diff
@@ -0,0 +1,45 @@
+     * @param {number} [minDecimals]
+     * @returns {string}
+     */
+    static formatNumber(value, maxDecimals = 5, minDecimals = Math.min(0, maxDecimals)) {
```
Why is the `Math.min` required for the `minDecimals`? Why is it not enough to just put `0` as the default? Or is it supposed to be `Math.max(0, minDecimals)` to not have negative decimals?

danimoh replied:
It's indeed not.
This was before `Math.min(2, maxDecimals))` but since I changed it to `0`, the `Math.min` is not required anymore.

sisou commented on src/lib/NumberFormatting.js:
```diff
@@ -0,0 +1,45 @@
+     * @param {string} [locale = navigator.language]
+     * @returns {string}
+     */
+    static formatCurrency(value, currency, locale = navigator.language) {
```
```suggestion
    static formatFiat(value, currency, locale = navigator.language) {
```

danimoh replied:
Technically, this could also be used to format other currencies like `BTC`, `NIM`, etc.

sisou replied:
Last I checked, `value.toLocaleString()` did not support crypto currencies :stuck_out_tongue: So it cannot, in fact, be used for crypto _currencies_ (https://www.currency-iso.org/dam/downloads/lists/list_one.xml).

danimoh replied:
Well, here is the specification for it: https://www.ecma-international.org/ecma-402/2.0/#sec-iswellformedcurrencycode

sisou replied:
OK, I have tested it and it actually accepts any three letter code and if it doesn't know it it just displays the uppercase three letters in front of the number formatted to (at least) two decimals.

OK fine.

sisou commented on src/components/PaymentInfoLine.js:
```diff
@@ -0,0 +1,78 @@
+
+        /** @type HTMLElement */
+        const $fiatAmount = (this.$el.querySelector('.fiat-amount'));
+        if (paymentInfo.fiatAmount !== undefined && paymentInfo.fiatCurrency !== undefined) {
```
Can `fiatAmount` also be `null` and would that be handled differently? Otherwise I think `fiatAmount` can not be `0` (should be validated in the RequestParser), thus we can use the simple truthy check here instead:

```suggestion
        if (paymentInfo.fiatAmount && paymentInfo.fiatCurrency) {
```

danimoh replied:
`fiatAmount` can be `number|undefined`. The check for `undefined` was just to make this explicit.
But I can change it.

sisou commented on src/components/PaymentInfoLine.js:
```diff
@@ -0,0 +1,78 @@
+
+        /** @type HTMLElement */
+        const $timer = (this.$el.querySelector('.timer'));
+        if (paymentInfo.time !== undefined && paymentInfo.expires !== undefined) {
```
I think both `time` and `expires` can also not be the number `0`, so it's safe to use truthy checks here:

```suggestion
        if (paymentInfo.time && paymentInfo.expires) {
```

danimoh replied:
Changed this too.

sisou commented on src/components/Timer.css:
```diff
@@ -0,0 +1,105 @@
+    color: white;
+    z-index: 1;
+    pointer-events: none;
+    background-image: url('data:image/svg+xml,<svg viewBox="0 0 136 63.9" xmlns="http://www.w3.org/2000/svg"><path d="M136 59.9a4 4 0 0 1-4 4H4a4 4 0 0 1-4-4v-47a4 4 0 0 1 4-4h99a4 4 0 0 0 3.2-1.7l4.6-6.6c.6-.8 1.8-.8 2.4 0l4.6 6.6a4 4 0 0 0 3.3 1.7H132a4 4 0 0 1 4 4z" fill="url(%23a)"/><defs><radialGradient id="a" cx="0" cy="0" r="1" gradientTransform="matrix(-190 0 0 -64.999 136 70.9)" gradient ...
```
Does the timer tooltip background need to be an SVG instead of `background-image: var(--nq-blue-bg)`?

danimoh replied:
Yes, the background image includes the small tooltip dent.

sisou commented on src/components/Timer.js:
```diff
@@ -0,0 +1,297 @@
+     * @returns {number}
+     */
+    get _timeLeft() {
+        return Math.max(0, Math.min(this._totalTime, this._endTime - Date.now()));
```
I think in the Hub we are working with offsets to the browser time to prevent mis-set browser clocks from breaking the timer. So when the page loads, the offset between the `startTime` and `Date.now()` is calculated and used in all relative calculations. Or is this offset already applied outside the timer?

danimoh replied:
The keyguard expects the times already to be converted to the local time.

sisou commented on src/lib/NumberFormatting.js:
```diff
@@ -0,0 +1,45 @@
+        const integerPart = integerMatch[0];
+        return formatted
+            .replace(/(\d)\D(\d)/, '$1.$2') // enforce a dot as decimal separator for
+            // consistency. Using capturing groups instead of lookahead/lookbehind to avoid browser support limitations.
```
I would be happier if this two-line comment is above the line that it's about.

danimoh replied:
Alright, then let's make you happy.

sisou commented on src/lib/Tweenable.js:
```diff
@@ -0,0 +1,68 @@
+     * @param {number} [startValue=targetValue]
+     * @param {number} [tweenTime=0]
+     * @param {number} [startTime=Date.now()]
+     * @param {(function(number): number)} [easing=Tweenable.Easing.EASE_IN_OUT_CUBIC]
```
_All_ arguments are optional?

Also, what is `tweenTime`? Can that get an explaining doc comment? Is that the _duration_?

danimoh replied:
Yes, it's the duration. Renamed it to `tweenDuration`.

sisou commented on src/request/sign-transaction/SignTransaction.js:
```diff
@@ -46,19 +49,37 @@ class SignTransaction {
+            imageUrl: 'shopLogoUrl' in request && !!request.shopLogoUrl
+                ? request.shopLogoUrl
+                : null,
             accountLabel: null,
```
I think for consistency the image assignment conditional should also be pulled above the object.

sisou commented on src/request/sign-transaction/SignTransaction.js:
```diff
@@ -102,6 +123,10 @@ class SignTransaction {
             },
         );
+
+        if ('expires' in request && request.expires !== undefined) {
```
Again, I think expires can not be `0`, so a truthy check is enough.

```suggestion
        if (request.expires) {
```

danimoh replied:
The `'expires' in request` is required for ts to know that this is a request on which `expires` exists.

sisou commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -23,14 +24,29 @@ class SignTransactionApi extends TopLevelApi {
             parsedRequest.shopOrigin = this.parseShopOrigin(request.shopOrigin);
             parsedRequest.shopLogoUrl = this.parseShopLogoUrl(request.shopLogoUrl);
             if (parsedRequest.shopLogoUrl && parsedRequest.shopLogoUrl.origin !== parsedRequest.shopOrigin) {
                 throw new Errors.InvalidRequestError('origin of shopLogoUrl must be same as referrer');
```
Since the referrer can't be checked in the Keyguard, this error message is technically wrong. And I think this is also checked in the Hub already? Why not:

```suggestion
                 throw new Errors.InvalidRequestError('origin of shopLogoUrl must be same as shopOrigin');
```

sisou commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -23,14 +24,29 @@ class SignTransactionApi extends TopLevelApi {
+
+            parsedRequest.time = this.parseNonNegativeFiniteNumber(request.time);
+            parsedRequest.expires = this.parseNonNegativeFiniteNumber(request.expires);
+            if (parsedRequest.expires !== undefined) {
+                if (parsedRequest.time === undefined) {
```
`time` and `expires` should never be `0`, right? So all these tests for `undefined` specifically are unnecessary IMO.

danimoh replied:
In this case typescript can't infer that in line 46 `parsedRequest.time` can be `undefined` if not checking for `undefined` here.
In other cases it's essentially a matter of taste.

sisou replied:
For me it's not just about taste, but also about defensive programming. Just because we tell Typescript that it can be `string | undefined` (for example), does not mean that outside calls cannot pass `null` or other types in. Typescript does not do type checks during runtime, and when it's not checked by whitelist anywhere, types are passed through as they are. So to be sure, I would always do a falsy or whitelist check for formerly untested arguments.

sisou commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -23,14 +24,29 @@ class SignTransactionApi extends TopLevelApi {
+            parsedRequest.expires = this.parseNonNegativeFiniteNumber(request.expires);
+            if (parsedRequest.expires !== undefined) {
+                if (parsedRequest.time === undefined) {
+                    throw new Errors.InvalidRequestError('If `expires` is defined `time` must be defined too.');
```
```suggestion
                    throw new Errors.InvalidRequestError('If `expires` is given, `time` must be given, too.');
```

sisou commented on tsconfig.json:
```diff
@@ -11,7 +11,7 @@
     // "outFile": "./",                       /* Concatenate and emit output to single file. */
     "outDir": "./.ts-out",                    /* Redirect output structure to the directory. */
-    "incremental": true,
+    "incremental": false,
```
What problem did you run into that you had to disable incremental compilation?

danimoh replied:
The build was using cached results that did not resemble what was in the code.
The non-incremental build is still very quick and for an app as important as the Keyguard it's worth to wait the extra time to get a proper build or rather typecheck.
The time difference is not really noticeable anyways.

sisou commented on types/Keyguard.d.ts:
```diff
@@ -54,15 +54,29 @@ type ConstructTransaction<T extends KeyguardRequest.TransactionInfo> = Transform
+        Transform<
+            ConstructTransaction<KeyId2KeyInfo<KeyguardRequest.SignTransactionRequestCheckout>>,
+            'shopLogoUrl',
+            { shopLogoUrl?: URL }
+        > :
+    T extends Is<T, KeyguardRequest.SignTransactionRequestSimple> ?
+        ConstructTransaction<KeyId2KeyInfo<KeyguardRequest.SignTransactionRequestSimple>>
         & { layout: KeyguardRequest.SignTransactionRequestLayout } :
-    T extends Is<T, KeyguardRequest.SignMessageRequest> ? Transform<KeyId2KeyInfo<KeyguardRequest.SignMessageRequest>,
-        'signer' | 'message', { signer: Nimiq.Address, message: Uint8Array | string }> :
+    T extends Is<T, KeyguardRequest.SignMessageRequest> ?
+        Transform<
+            KeyId2KeyInfo<KeyguardRequest.SignMessageRequest>,
+            'signer' | 'message',
+            { signer: Nimiq.Address, message: Uint8Array | string }
+        > :
     T extends Is<T, KeyguardRequest.SimpleRequest>
         | Is<T, KeyguardRequest.DeriveAddressRequest>
         | Is<T, KeyguardRequest.RemoveKeyRequest>
         | Is<T, KeyguardRequest.ExportRequest> ? KeyId2KeyInfo<T> :
-    T extends Is<T, KeyguardRequest.ImportRequest> ? Transform<KeyguardRequest.ImportRequest,
-        'isKeyLost', { isKeyLost: boolean }> :
+    T extends Is<T, KeyguardRequest.ImportRequest> ?
+        Transform<
+            KeyguardRequest.ImportRequest,
+            'isKeyLost',
+            { isKeyLost: boolean }
+        > :
```
Not really more readable IMO.

danimoh replied:
In my opinion it is.

sisou commented:
I would appreciate it if updates to this PR are done as commits on top of the existing commits, and _not_ as rebases ;). Makes reviewing the updates much easier.

My requested changes are mostly small, simple stuff, that I nontheless think will make readability and understanding easier for devs unfamiliar with the code.

danimoh committed:
Small code style changes in checkout related code

danimoh commented:
Well, reviewing force-pushed changes is not really much harder. You just have to diff from the commit hash that you last reviewed at (which github tells you) to the latest commit. All this can be done in github too.
